### PR TITLE
Grow `RawVec` to fill the allocator bins tighter [please bench it]

### PR DESCRIPTION
**Pull request by dpc**
AFAIK, most if not not all memory allocators use some combination
of evenly sized bins. Typically these bins are power of two sized,
sometimes suplemented with some 0b1100xxx... bins as well.

Most of the time every allocation is also prefixed with a pointer to some
per-allocation metadata, adding a fixed overhead to every requested allocation.

This can observed with:

```rust
fn main() {
    let s = 24;
    let v1: Vec<u8> = Vec::with_capacity(s);
    let v2: Vec<u8> = Vec::with_capacity(s);
    let v3: Vec<u8> = Vec::with_capacity(s);

    println!("{:?}", v1.as_ptr());
    println!("{:?}", v2.as_ptr());
    println!("{:?}", v3.as_ptr());
}
```

https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=d091d5b6fc0ea89e5e07438bdb740ac3

For `let s = 24` the pointers are 32 bytes apart, but increasing `s` to 25 make
 them 64 bytes apart, as the allocation falls into the one up sized bin.

This made me think that the default way of growing collections in Rust
(doubling their capacity) is degenerate in most common cases. ~~Most types
is aligned to the power of 2 size, and then doubling their
size make almost every allocation waste almost 50% of actually allocated
space for it.~~

**Edit** After a bit more experimenting: For larger sizes (> 64),
the allocator returns pointers with 16 of extra space between.
Maybe these are some skip lists, maybe something else. I crossed
out the naive conclusion, that works only for small allocations.
However the point stands. Whatever the allocator is using is most probably
not the optimal case. By trying to allocate capacity to the optimal value,
we can opportunistically save memory and make allocator's work easier.

By growing the capacity by trying to fill the bin well, we can possibly
avoid some needless allocations *and* lower the memory consumption.

TODOs:

* [ ] get to compile & pass tests
* [ ] do initial bench
* [x] don't do `2*cap` anymore
* [ ] put a cap on how big of an allocation this is(?); for bigger sizes allocator might be doing something else than bins; what cutoff to use thought?

### CODE REVIEW

**Comment by leonardo-m** [comment]
I'm curious to see benchmarks of this :)

File: library/alloc/src/raw_vec.rs [hunk 1]
```diff
@@ -394,6 +394,24 @@ impl<T, A: Allocator> RawVec<T, A> {
...
+
+        // Since memory allocators tend to use power of two sized bins, find the
+        // bin size we will fall into.
+        let bin_size = usize::MAX >> (alloc_size.leading_zeros() - 1);
+
+        // Leave some room for allocators that add fixed overhead (usually one pointer-size)
+        // FIXME: maybe two pointers to be on the safe side? It could potentially be platform-dependent.
+        let aligned_alloc_size = bin_size.saturating_sub(mem::align_of::<usize>());
+
+        // Ignore cases where the value is actually smaller, due to substractiing fixed overhead
```
Comment (scottmcm) [inline library/alloc/src/raw_vec.rs:419]: This makes me notice that the current approach is basically
```rust
let new_cap = max![ len + additional, cap * 2, MIN_CAP ];
```
so I wonder if it might make sense to just make this another thing in that list, or even replacing the `cap * 2`.

`sizeof(usize).div_ceil(sizeof(T))` could be a constant like `MIN_NON_ZERO_CAP` is, so it's easy to subtract off even in length units instead of needing to switch to byte units here.

File: library/alloc/src/raw_vec.rs [hunk 2]
```diff
@@ -394,6 +394,24 @@ impl<T, A: Allocator> RawVec<T, A> {
...
         let cap = cmp::max(Self::MIN_NON_ZERO_CAP, cap);
 
+        let alloc_size = cap.checked_mul(mem::align_of::<T>()).ok_or(CapacityOverflow)?;
+
+        // Since memory allocators tend to use power of two sized bins, find the
+        // bin size we will fall into.
+        let bin_size = usize::MAX >> (alloc_size.leading_zeros() - 1);
+
+        // Leave some room for allocators that add fixed overhead (usually one pointer-size)
+        // FIXME: maybe two pointers to be on the safe side? It could potentially be platform-dependent.
```
Comment (scottmcm) [inline library/alloc/src/raw_vec.rs:416]: Hmm, this is a good point.  This change would massively backfire if it underestimates the overhead, as the underlying allocator would then need to massively overallocate basically every time, instead of the sometimes you'd get today depending on the original capacity.

So maybe this should be very conservative in the space it leaves for the allocator overhead?

Comment (scottmcm) [reply library/alloc/src/raw_vec.rs:416]: Thinking about it again, though, is leaving any space at all pessimizing good allocators in favour of mediocre ones?

One big advantage of rust's allocator APIs is that deallocation passes the size.  So, for example, if an allocator allocates dedicated (maybe-large) pages for 32-byte allocations, they don't necessarily need any per-allocation overhead to do that, as they can tell by the deallocation size that it's a page of 32-byte allocations, and could have per-page overhead instead.

**Review by scottmcm — COMMENTED** [review]
meta-comment: Maybe it'd be worth adjusting this by the size of the allocation too?  For an allocation of a few dozen megabytes, it's not so obvious the rounding up to a power of two is necessary, as they might just get pages from the OS for the actual size, rather than bucketing.

File: library/alloc/src/raw_vec.rs [hunk 3]
```diff
@@ -394,6 +394,24 @@ impl<T, A: Allocator> RawVec<T, A> {
         let cap = cmp::max(self.cap * 2, required_cap);
```
[the remark this answers is not in the document]
Comment (dpc) [reply library/alloc/src/raw_vec.rs:402]: Agreed.

**Comment by dpc** [comment]
I fixed the `clz` usage (I think) and pushed that.

I got excited about this change, but I'm running out of time today, and will be busy over the weekend. If anyone feels like working on it in the meantime, please feel free. I'm very curious if this theory leads to any noticeable improvements.

File: library/alloc/src/raw_vec.rs (hunk 2 above)
Comment (dpc) [reply library/alloc/src/raw_vec.rs:416]: Generally, I think that any allocator that uses most than just a single pointer is meh, and is not worth pessimizing. 

But yeah... unfortunately this optimization is assuming implementation details that are invisible. Hopefully benchmarks will tell if any platform degraded, and we could always use conditional compilation or something to tweak it.

**Comment by CAD97** [comment]
> Most types is aligned to the power of 2 size

*All* types are aligned to a power of two. This is a hard requirement of Rust types.

-----

It's worth noting that allocations through `Allocator` return a layout for exactly this purpose. The "proper" way to do this imho isn't to do this logic in RawVec, but to instead use the extra space which the allocator reports. That way, you match the specifics of the allocator in use directly.

However, the current common allocators (system and jemalloc) don't actually return this information, because it's typically extra overhead to get that dynamic information from the allocator, if it's available at all. It's also a bit more complicated than just requesting a well-fit size, because the allocator has to accept any layout for freeing between the requested and reported provided layouts. If the underlying allocator implementation doesn't support this, even if it would benefit from better bin fitting, the Rust allocator wrapper can't do such adjustments.

File: library/alloc/src/raw_vec.rs (hunk 2 above)
Comment (CAD97) [reply library/alloc/src/raw_vec.rs:416]: There's another interesting case to consider: the windows `System` allocator has to implement alignment on top of allocation which does not support custom alignment. This means it allocates `align` extra bytes.

Yes, Rust's allocation API allows the use of allocators which don't have to store metadata in allocations, but in practice, most allocators will want to support the C allocation API.

Comment (dpc) [reply library/alloc/src/raw_vec.rs:408]: > alignment on top of allocation which does not support custom alignment. This means it allocates align extra bytes.

@CAD97 I don't understand.

Comment (CAD97) [reply library/alloc/src/raw_vec.rs:408]: https://github.com/rust-lang/rust/blob/dec689432fac6720b2f18101ac28a21add98b1b8/library/std/src/sys/windows/alloc.rs#L129-L178

Roughly:

```rust
fn win32::HeapAlloc(hHeap: win32::HANDLE, dwFlags: win32::DWORD, dwBytes: win32::SIZE_T) -> win32::LPVOID;
fn win32::HeapFree (hHeam: win32::HANDLE, dwFlags: win32::DWORD, lpMem: win32::LPVOID);

fn HeapAlignedAlloc(heap: win32::HANDLE, flags: win32::DWORD, size: usize, align: usize) -> *mut u8 {
    if align <= win32::MEMORY_ALLOCATION_ALIGNMENT {
        return win32::HeapAlloc(heap, flags, size) as *mut u8;
    }
    let ptr = win32::HeapAlloc(heap, flags, size + align);
    let aligned = (ptr as *mut u8).add(1).align_to(align);
    *(aligned as *mut win32::LPVOID).sub(1) = ptr;
    aligned
}

fn HeapAlignedFree(heap: win32::HANDLE, flags: win32::DWORD, ptr: *mut u8, size: usize, align: usize) {
    if align <= win32::MEMORY_ALLOCATION_ALIGNMENT {
        win32::HeapFree(heap, flags, ptr as _)
    } else {
        win32::HeapFree(heap, flags, *(ptr as *mut win32::LPVOID).sub(1))
    }
}
```

Windows doesn't provide a way to `HeapAlloc` at anything other than `MEMORY_ALLOCATION_ALIGNMENT`. (Windows does offer `_alligned_malloc`, `_aligned_realloc`, `_aligned_recalloc`, and `_aligned_free`, but strangely not `_aligned_calloc`.) `std::alloc::System` is documented to use `HeapAlloc` on Windows.

Comment (dpc) [reply library/alloc/src/raw_vec.rs:408]: > There's another interesting case to consider: the windows `System` allocator has to implement alignment on top of allocation which does not support custom alignment. This means it allocates `align` extra bytes.

Wouldn't memory allocator request much, much larger chunks of memory from the system at once, and then use them for fine-grained user-allocations individually? So any system level alignment issues would generally have much smaller overhead?

**Comment by dpc** [comment]
How will I bench it once I finally get it to even work? OK to point me to links, I just failed to find it.

**Comment by CAD97** [comment]
A usual [rust-timer](https://perf.rust-lang.org/help.html) run is probably sufficient to see the performance impact. Whoever reviews will kick that off.

File: library/alloc/src/raw_vec.rs (hunk 2 above)
Comment (CAD97) [reply library/alloc/src/raw_vec.rs:408]: An actual new allocator implementation, yes, typically. I'm specifically referring to `System` or other Rust wrappers of allocators which don't provide an aligned allocation interface, which need to do this technique on a per-allocation basis.

While this may be only the Windows system allocator in practice, this is still the default allocator that most Rust programs compiled for Windows will be using, so it's an important case to consider.

File: library/alloc/src/raw_vec.rs [hunk 4]
```diff
@@ -394,6 +394,28 @@ impl<T, A: Allocator> RawVec<T, A> {
...
         let cap = cmp::max(Self::MIN_NON_ZERO_CAP, cap);
 
+        let alloc_size = cap.checked_mul(mem::align_of::<T>()).ok_or(CapacityOverflow)?;
```
Comment (dpc) [inline library/alloc/src/raw_vec.rs:397]: This incorrect, which became clear when `[u8; n]` tests started to run. I'll fix to use `size_of`. Please someone correct me if `size_of` does not already include alignment requirements.

File: library/alloc/src/raw_vec.rs (hunk 1 above)
Comment (CAD97) [reply library/alloc/src/raw_vec.rs:410]: > instead of needing to switch to byte units here

The byte units are used because what we're targeting is a byte size of (just under) a power of two. We can `const` provide "`usize` per `T`" and "`T` per `usize`", but I don't see how to calculate `(size_of::<T>() * cap).next_power_of_two() / size_of::<T>()` without scaling `cap` into byte units instead of `T` units.

File: library/alloc/src/raw_vec.rs (hunk 4 above)
Comment (CAD97) [reply library/alloc/src/raw_vec.rs]: Rust guarantees that size is a multiple of align.

File: library/alloc/src/raw_vec.rs (hunk 2 above)
Comment (dpc) [reply library/alloc/src/raw_vec.rs:407]: I kind of understand, but not enough to come up with a solution. Worst case we could just have a different implementation for Windows maybe?

**Comment by dpc** [comment]
OK. Basic version passed, now I did some tweaks on top of just adding up sizing.

File: library/alloc/src/raw_vec.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (dpc) [reply library/alloc/src/raw_vec.rs]: Done.

**Comment by dpc** [comment]
> meta-comment: Maybe it'd be worth adjusting this by the size of the allocation too? For an allocation of a few dozen megabytes, it's not so obvious the rounding up to a power of two is necessary, as they might just get pages from the OS for the actual size, rather than bucketing.

I agreed initially, but now I'm thinking it's not going to be worse than what normal `2x`-ing was doing before, no?

**Comment by Mark-Simulacrum** [comment]
FWIW, some prior art here: https://github.com/rust-lang/rust/pull/45482, https://github.com/rust-lang/rust/pull/50739, https://github.com/rust-lang/rust/pull/29848, https://github.com/rust-lang/rust/pull/29454, https://github.com/rust-lang/rust/pull/72227

@bors try @rust-timer queue -- might as well kick off some perf benchmarks as well, I suppose.

**Comment by dpc** [comment]
*oof*

File: library/alloc/src/raw_vec.rs [hunk 5]
```diff
@@ -389,10 +392,31 @@ impl<T, A: Allocator> RawVec<T, A> {
...
         let required_cap = len.checked_add(additional).ok_or(CapacityOverflow)?;
 
-        // This guarantees exponential growth. The doubling cannot overflow
-        // because `cap <= isize::MAX` and the type of `cap` is `usize`.
-        let cap = cmp::max(self.cap * 2, required_cap);
-        let cap = cmp::max(Self::MIN_NON_ZERO_CAP, cap);
+        // A minimum fixed cap increase along with sizing up to the allocator bin
+        // size guarantees exponential growth.
+        // The addition can't overflow because `cap <= isize::MAX` and the type of `cap` is `usize`.
+        let cap = cmp::max(self.cap + 2 * mem::size_of::<usize>(), required_cap);
```
Comment (dpc) [inline library/alloc/src/raw_vec.rs:398]: BUG: This is in capacity, not bytes.

File: library/alloc/src/raw_vec.rs [hunk 6]
```diff
@@ -389,10 +392,31 @@ impl<T, A: Allocator> RawVec<T, A> {
...
+        // Leave some room for allocators that add fixed overhead (usually
+        // one pointer-size)
+        // FIXME: maybe two pointers to be on the safe side? It could potentially
+        // be platform-dependent.
+        let aligned_alloc_size = bin_size.saturating_sub(mem::align_of::<usize>() - 1) /* the +1 skipped from the previous line turned into -1 here */ ;
+
+        // Ignore cases where the value is actually smaller, due to substractiing the fixed overhead
+        let cap = if alloc_size < aligned_alloc_size {
+            // Align the capacity to fit the bin
+            aligned_alloc_size / mem::align_of::<T>()
```
Comment (dpc) [inline library/alloc/src/raw_vec.rs:416]: BUG: This was supposed to be `size_of`, not `align_of`.

**Comment by dpc** [comment]
> FWIW, some prior art here: #45482, #50739, #29848, #29454, #72227
> 
> @bors try @rust-timer queue -- might as well kick off some perf benchmarks as well, I suppose.

@Mark-Simulacrum 
Results were so bad that there had to be some glaring bugs. I've figured out how to run tests locally and I've found two big bugs and fixed them. Now I have a sanity tests as well. :crossed_fingers: 

If there's a way to run these (or some) benchmarks locally, I would be happy to know, as it would help me experiment. However right now I'm hoping to get benchmark results close to the baseline one way or another.

File: library/alloc/src/raw_vec.rs [hunk 7]
```diff
@@ -386,13 +388,32 @@ impl<T, A: Allocator> RawVec<T, A> {
...
+        let alloc_overhead_size = mem::size_of::<usize>();
+
         // Nothing we can really do about these checks, sadly.
         let required_cap = len.checked_add(additional).ok_or(CapacityOverflow)?;
 
-        // This guarantees exponential growth. The doubling cannot overflow
-        // because `cap <= isize::MAX` and the type of `cap` is `usize`.
-        let cap = cmp::max(self.cap * 2, required_cap);
-        let cap = cmp::max(Self::MIN_NON_ZERO_CAP, cap);
+        let alloc_size = required_cap.checked_mul(mem::size_of::<T>()).ok_or(CapacityOverflow)?;
```
Comment (dpc) [inline library/alloc/src/raw_vec.rs:399]: All these new checked operations are unfortunate, but I'll worry about them only if I get any improvements in the RSS that validates the idea.

**Comment by Mark-Simulacrum** [comment]
@bors try @rust-timer queue

See https://github.com/rust-lang/rustc-perf/tree/master/collector#how-to-benchmark-a-change-on-your-own-machine (and other content in that file is maybe useful).

**Comment by dpc** [comment]
I spoke too soon and even the basic tests didn't pass yet. Took me another hour to fix some corner cases. We'll see if everything is green now. The pain is real.

Also the code accumulated a lot checks if conditions, so I'm growing pessimistic.

**Comment by dpc** [comment]
@Mark-Simulacrum Done, CI green, sanity test is there. Not sure how it works, an if I ruined this one run. Please let's bench it.

In the meantime I'm trying to get the setup you linked to working, but I have some issues as I'm on NixOS.

**Comment by dpc** [comment]
Hey, the RSS looks OK, if I'm reading it right. If only I could optimize that code, maybe it would be a win...

**Comment by dpc** [comment]
OK, I pushed a simple change that should make it allocate more at once (on average). I'm still setting up things to be able to bench locally.

@Mark-Simulacrum Please try again. Thank you.

**Comment by dpc** [comment]
Is it still in a queue, or something bad happened to it?

**Comment by Kobzol** [comment]
Still in queue, it was stuck for some time today. You can see the status at perf.rust-lang.org/status.html.

**Comment by dpc** [comment]
One important observation: this optimization doesn't make much sense for bigger `T`s as it's unlikely that they can be aligned enough for this to make a difference. I've added a change to address it.

However I figured out how to bench it locally, an the results are not great. Generally small slowdown across the board, for no trend of even any memory savings. Handful of binaries seem to benefit (-8%), but some other handful suffer (+10%). 

I suspect that a lot of binaries actually has a need for exactly "2^n" sized buffers, in which case my change just gets in a way by over-sizing the capacity for them to to the fit with overhead to next power of two sized bin.

File: library/alloc/src/raw_vec.rs (hunk 1 above)
Comment (scottmcm) [reply library/alloc/src/raw_vec.rs]: Oh, right!  Good point.

**Comment by nnethercote** [comment]
A warning: I have spent quite some time working on the `RawVec` growth code in the past. It is very hot and even tiny changes can have surprisingly large performance effects. The existing code has been tuned extensively, and numerous attempts at improvements have failed. That's not to say there is no room for improvements, but I think the low-hanging fruit has already been picked here.

**Comment by dpc** [comment]
It was a fun time, but I think it's time to give up. I pushed the best I was able to get: slight memory decrease in some cases (majority), slight decrease in some others (minority, but seems like some of the most "significant" ones). On top of it the instruction/cycles were slightly negative across the board.

Seems that allocating to `n^2 - sizeof::<usize>()` does kind of save memory when you end up with something under under that capacity. My intuition tells me that what ruins it is that a lot of software does use a lot of  2^n sized buffers, and in this case rounding up leads to almost 50% over-allocated buffer.

On top of it calculating all the sizes etc. is additional overhead, and in has to deal with overflow-checking math.
